pkg/envtest/webhook.go

```diff
@@ -0,0 +1,358 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	admissionreg "k8s.io/api/admissionregistration/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	kclient "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/internal/testing/integration"
+	"sigs.k8s.io/controller-runtime/pkg/internal/testing/integration/addr"
+	"sigs.k8s.io/yaml"
+)
+
+// WebhookInstallOptions are the options for installing mutating or validating webhooks
+type WebhookInstallOptions struct {
+	// Paths is a list of paths to the directories containing the mutating or validating webhooks yaml or json configs.
+	DirectoryPaths []string
+
+	// MutatingWebhooks is a list of MutatingWebhookConfigurations to install
+	MutatingWebhooks []*admissionreg.MutatingWebhookConfiguration
+
+	// ValidatingWebhooks is a list of ValidatingWebhookConfigurations to install
+	ValidatingWebhooks []*admissionreg.ValidatingWebhookConfiguration
+
+	// ErrorIfPathMissing will cause an error if a DirectoryPath does not exist
+	ErrorIfPathMissing bool
+
+	// LocalServingHost is the host for serving webhooks on.
+	// it will be automatically populated
+	LocalServingHost string
+
+	// LocalServingPort is the allocated port for serving webhooks on.
+	// it will be automatically populated by a random available local port
+	LocalServingPort int
+
+	// LocalServingCertDir is the allocated directory for serving certificates.
+	// it will be automatically populated by the local temp dir
+	LocalServingCertDir string
+
+	// MaxTime is the max time to wait
+	MaxTime time.Duration
+
+	// PollInterval is the interval to check
+	PollInterval time.Duration
+}
+
+// ModifyWebhookDefinitions modifies webhook definitions by:
+// - applying CABundle based on the provided tinyca
+// - if webhook client config uses service spec, it's removed and replaced with direct url
+func (o *WebhookInstallOptions) ModifyWebhookDefinitions(caData []byte) error {
+	hostPort, err := o.generateHostPort()
+	if err != nil {
+		return err
+	}
+
+	for _, hookSet := range o.MutatingWebhooks {
+		for i := range hookSet.Webhooks {
+			cfg := &hookSet.Webhooks[i].ClientConfig
+			cfg.CABundle = caData
+			var path string
+			if cfg.Service != nil && cfg.Service.Path != nil {
+				path = *cfg.Service.Path
+			}
+			cfg.Service = nil // we cannot use service in integration tests since we're running controller outside cluster
+			url := fmt.Sprintf("https://%s/%s", hostPort, path)
+			cfg.URL = &url
+		}
+	}
+	for _, hookSet := range o.ValidatingWebhooks {
+		for i := range hookSet.Webhooks {
+			cfg := &hookSet.Webhooks[i].ClientConfig
+			cfg.CABundle = caData
+			var path string
+			if cfg.Service != nil && cfg.Service.Path != nil {
+				path = *cfg.Service.Path
+			}
+			cfg.Service = nil // we cannot use service in integration tests since we're running controller outside cluster
```

Comment:
with this the comment, is the intent here to generate a panic if it's used incorrectly? what happens in existing tests if this is not nil but is invalid (and if invalid, are we missing certain other validations?)

Reply:
the intent here is that we swap out service for raw address because we don't have an actually standard kube service network.  We want to you to be able to use your standard config though (which probably has a service set, b/c that's how you deploy in prod), so we just blank it out here.

Reply:
I.e. you specify

```yaml
clientConfig:
  service:
    path: /foo
    name: somesvc
    namespace: somens
```

and we effectively convert internally to

```yaml
clientConfig:
  url: https://localhost:<some-random-port>/foo
```

so we can use it.

Reply:
yep, I made the comment more verbose to make this more obvious